### PR TITLE
Refactor newton solver scheme.

### PR DESCRIPTION
**Pull request by MFraters**
This is one of the first step towards the goal set in #3102. It puts the most of the newton solver scheme code in a seperate function.

### CODE REVIEW

**Comment by MFraters** [comment]
just so that it is clear, this is ready for review.

File: include/aspect/simulator.h [hunk 1]
```diff
@@ -581,6 +593,11 @@ namespace aspect
        */
       void solve_iterated_advection_and_newton_stokes ();
 
+      /**
+       * todo
```
Comment (gassmoeller) [inline include/aspect/simulator.h]: Todo

File: include/aspect/simulator.h [hunk 2]
```diff
@@ -581,6 +593,11 @@ namespace aspect
...
+       */
+      void assemble_and_solve_defect_correction_Stokes(DefectCorrectionResiduals &dcr, bool use_picard);
```
Comment (gassmoeller) [inline include/aspect/simulator.h]:
```suggestion
      void assemble_and_solve_defect_correction_Stokes(DefectCorrectionResiduals &dcr, 
                                                       const bool use_picard);
```

File: source/simulator/solver_schemes.cc [hunk 3]
```diff
@@ -565,311 +566,314 @@ namespace aspect
...
+  }
 
-        // the Stokes matrix depends on the viscosity. if the viscosity
-        // depends on other solution variables, then after we need to
-        // update the Stokes matrix in every time step and so need to set
-        // the following flag. if we change the Stokes matrix we also
-        // need to update the Stokes preconditioner.
-        rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+  template <int dim>
+  void Simulator<dim>::assemble_and_solve_defect_correction_Stokes(DefectCorrectionResiduals &dcr, bool use_picard)
```
Comment (gassmoeller) [inline source/simulator/solver_schemes.cc]: Please move this up to the other `assemble_and_solve_...` functions.

**Comment by MFraters** [comment]
I think I addressed all the comments.

**Comment by MFraters** [comment]
@gassmoeller could you take an other look?

File: include/aspect/simulator.h [hunk 4]
```diff
@@ -718,6 +730,20 @@ namespace aspect
       double assemble_and_solve_stokes (const bool compute_initial_residual = false,
                                         double *initial_nonlinear_residual = nullptr);
 
+      /**
+       * Assemble and solve the defect correction for of the Stokes equation.
```
Comment (gassmoeller) [inline include/aspect/simulator.h]:
```suggestion
       * Assemble and solve the defect correction form of the Stokes equation.
```

File: include/aspect/simulator.h [hunk 5]
```diff
@@ -718,6 +730,20 @@ namespace aspect
...
+       * This function takes a structure of DefectCorrectionResiduals which
+       * contains information about different resuduals. The information in
```
Comment (gassmoeller) [inline include/aspect/simulator.h]:
```suggestion
       * contains information about different residuals. The information in
```

File: include/aspect/simulator.h [hunk 6]
```diff
@@ -718,6 +730,20 @@ namespace aspect
...
+       * this structure is updated by this function. The parameter use_picard
+       * forces the use of the defect correction Piccard iteration, so no
```
Comment (gassmoeller) [inline include/aspect/simulator.h]:
```suggestion
       * forces the use of the defect correction Picard iteration, if set to 'true' no
```

**Comment by MFraters** [comment]
> Looks good. This PR moves a lot of code from `solve_iterated_advection_and_newton_stokes` into a new function. Is this function now usable instead of the code in `solve_single_advection_iterated_newton_stokes`? This would save some 100-200 lines already. If not let me know, this by itself is already a nice step forward.

Thanks for spotting the spelling mistakes :) Yes, that was added in between that this pull request was created and now. Will make a separate pull request for that.

**Pull request merged**